lib/core/checker/checker.cc

```diff
@@ -10,6 +19,263 @@
 
 namespace dbuf::checker {
 
+struct NameResolutionChecker {
+  template <typename T>
+  bool operator()(const T &) const {
+    return false;
+  }
+
+  bool operator()(const ast::AST &ast) {
+    AddScope();
+    AddGlobalNames(ast);
+
+    constructors_fields_ = GetConstructorFields(ast);
+
+    if (!(*this)(ast.messages)) {
+      return false;
+    }
+
+    return (*this)(ast.enums);
+  }
+
+  bool operator()(const std::unordered_map<InternedString, ast::Enum> &enums) {
+    return !std::ranges::any_of(enums.begin(), enums.end(), [this](auto &ast_enum) {
+      return !(*this)(ast_enum.second);
+    });
+  }
+
+  bool operator()(const ast::Enum &ast_enum) {
+    AddScope();
+    if (!IsInScope(ast_enum.name)) {
+      return false;
+    }
+    if (!(*this)(ast_enum.type_dependencies)) {
+      return false;
+    }
+    if (!(*this)(ast_enum.pattern_mapping)) {
+      return false;
+    }
+    DeleteScope();
+
+    return true;
+  }
+
+  bool operator()(const std::vector<ast::Enum::Rule> &rules) {
+    return !std::ranges::any_of(rules.begin(), rules.end(), [this](auto &rule) {
+      return !(*this)(rule);
+    });
+  }
+
+  bool operator()(const ast::Enum::Rule &rule) {
+    AddScope();
+
+    isShadowing_ = true;
+
+    if (!(*this)(rule.inputs)) {
+      return false;
+    }
+
+    isShadowing_ = false;
+
+    if (!(*this)(rule.outputs)) {
+      return false;
+    }
+
+    DeleteScope();
+
+    return true;
+  }
+
+  bool operator()(std::vector<ast::Constructor> &constructors) {
+    return !std::ranges::any_of(
+        constructors.begin(),
+        constructors.end(),
+        [this](auto &constructor) { return !(*this)(constructor); });
+  }
+
+  bool operator()(const ast::Constructor &constructor) {
+    return (!IsInScope(constructor.name)) && ((*this)(constructor.fields));
+  }
+
+  bool operator()(const ast::Star &) {
+    return true;
+  }
+
+  bool operator()(const std::unordered_map<InternedString, ast::Message> &messages) {
+    return !std::ranges::any_of(messages.begin(), messages.end(), [this](auto &message) {
+      return !(*this)(message.second);
+    });
+  }
```

Comment:
These negations need some comments for me to grasp what's going on, I'm sorry :laughing: